### PR TITLE
feat(controller): Add podMetadata field to workflow spec. Resolves #4985

### LINKED ISSUE #4985: Default annotations for workflow pods
# Summary

It would be useful to have the ability to set default annotations for all pods in a workflow via the default workflow spec. Perhaps pods in a workflow could inherit the annotations on the workflow itself? Or another configuration option could be added for this purpose.

# Use Cases

We use kube2iam for AWS access which includes our s3 artifact repository. Unless I'm missing something, it seems like every template needs to have an annotation individually applied to it so that it can archive logs. I'd like to be able to apply a default annotation to all workflow pods via the default workflow spec, which would save every workflow definitions from needing to know about that role.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by dylanhellems**
Closes https://github.com/argoproj/argo/issues/4985

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -615,6 +615,14 @@ func isResourcesSpecified(ctr *apiv1.Container) bool {
 
 // addMetadata applies metadata specified in the template
 func (woc *wfOperationCtx) addMetadata(pod *apiv1.Pod, tmpl *wfv1.Template, opts *createWorkflowPodOpts) {
+	// inherit workflow annotations and labels
+	for k, v := range woc.wf.ObjectMeta.Annotations {
```
Comment (alexec) [inline workflow/controller/workflowpod.go:619]: I'm not sure we just want to copy the workflows annotations and labels to the pods. Instead, perhaps a new spec.podMetadata field might be better - @jessesuen ?

Comment (terrytangyuan) [reply workflow/controller/workflowpod.go:619]: I agree. We probably don't want to inherit directly. Perhaps we could add something similar to podSpecPatch, e.g. podMetadataPatch.

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -615,6 +615,14 @@ func isResourcesSpecified(ctr *apiv1.Container) bool {
...
+		pod.ObjectMeta.Annotations[k] = v
```
Comment (terrytangyuan) [inline workflow/controller/workflowpod.go:620]: I agree. We probably don't want to inherit directly as we don't know for sure whether to apply the same set of annotations/labels. Perhaps we could add something similar to podSpecPatch, e.g. podMetadataPatch.

**Comment by dylanhellems** [comment]
@alexec @terrytangyuan Makes sense; I've added a podMetadata field to the workflow spec. I have a couple of questions though:
1. I had to update the maxWorkflowSize in the packer and hydrator tests. Is that an arbitrary value for testing or are implications to that change that I'm unaware of?
2. I'm having trouble getting codegen working locally. Here is the output I'm seeing:
```
dylanhellems@Dylans-MBP argo % make codegen                                                        
# protoc pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto
[ -e vendor ] || go mod vendor
protoc -I /usr/local/include -I /Users/dylanhellems/Documents/repos/argo -I /Users/dylanhellems/Documents/repos/argo/vendor -I /Users/dylanhellems/go/src -I /Users/dylanhellems/go/pkg/mod/github.com/gogo/protobuf@v1.3.1/gogoproto -I /Users/dylanhellems/go/pkg/mod/github.com/grpc-ecosystem/grpc-gateway@v1.16.0/third_party/googleapis --gogofast_out=plugins=grpc:/Users/dylanhellems/go/src --grpc-gateway_out=logtostderr=true:/Users/dylanhellems/go/src --swagger_out=logtostderr=true,fqn_for_swagger_name=true:. pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto
github.com/argoproj/argo/pkg/apis/workflow/v1alpha1/generated.proto: File not found.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:6:1: Import "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1/generated.proto" was not found or had errors.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:15:5: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:31:5: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:41:5: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:46:87: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:53:81: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:57:84: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplateList" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:61:87: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.proto:72:83: "github.com.argoproj.argo.v3.pkg.apis.workflow.v1alpha1.ClusterWorkflowTemplate" is not defined.
make: *** [pkg/apiclient/clusterworkflowtemplate/cluster-workflow-template.swagger.json] Error 1
```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -351,6 +351,9 @@ type WorkflowSpec struct {
 
 	// RetryStrategy for all templates in the workflow.
 	RetryStrategy *RetryStrategy `json:"retryStrategy,omitempty" protobuf:"bytes,37,opt,name=retryStrategy"`
+
+	// PodMetadata defines additional metadata that should be applied to workflow pods
+	PodMetadata Metadata `json:"podMetadata,omitempty" protobuf:"bytes,38,opt,name=podMetadata"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:356]: should be pointer type

**Comment by dylanhellems** [comment]
Codegen issue above was because I wasn't working within the gopath. Working now.

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 3 above)
Comment (dylanhellems) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Thanks!

**Review by jessesuen — APPROVED** [review]

**Comment by alexec** [comment]
Please sync with master.

**Review by alexec — CHANGES_REQUESTED** [review]
Please resolve conflicts.

**Comment by dylanhellems** [comment]
@alexec Conflicts resolved

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
LGTM

**Pull request merged**